libsql-server/src/database.rs

```diff
@@ -35,14 +47,26 @@ pub struct PrimaryDatabase {
     pub connection_maker: Arc<dyn MakeConnection<Connection = PrimaryConnection>>,
 }
 
+#[async_trait]
 impl Database for PrimaryDatabase {
     type Connection = PrimaryConnection;
 
     fn connection_maker(&self) -> Arc<dyn MakeConnection<Connection = Self::Connection>> {
         self.connection_maker.clone()
     }
 
-    fn shutdown(&self) {
+    fn destroy(self) {
+        self.logger.closed_signal.send_replace(true);
+    }
+
+    async fn shutdown(self) -> Result<()> {
         self.logger.closed_signal.send_replace(true);
+        if let Some(replicator) = &self.logger.bottomless_replicator {
+            let replicator = replicator.lock().unwrap().take();
+            if let Some(mut replicator) = replicator {
+                replicator.wait_until_snapshotted().await?;
+            }
+        }
+        Ok(())
```

Comment:
Is this cancel safe? @MarinPostma do we have shutdown timeouts? What happens if this gets canceled? Do we have a half backed up backup? This is just waiting so I assume the actual backup happens on a bg thread that can't get cancelled while doing IO by tokio?

Reply:
@Horusiath ☝️

Reply:
afaik, there are recovery mecanisms when we detect primary is more up to date than bottomless

Reply:
> Is this cancel safe? @MarinPostma do we have shutdown timeouts? What happens if this gets canceled? Do we have a half backed up backup? This is just waiting so I assume the actual backup happens on a bg thread that can't get cancelled while doing IO by tokio?

In unusual circumstances, it can get cancelled in the middle of a shutdown.. may be cos of OOM or from Fly. I had tested this scenario earlier, IIRC, when the machine reboots, restoration from the last back up fails. This is easy to test via Nemesis, I will do that and report back.